tcping.go

```diff
@@ -19,18 +17,72 @@ import (
 	"github.com/google/go-github/v45/github"
 )
 
+// printer is a set of methods for printers to implement.
+//
+// Printers should NOT modify any existing data nor do any calculations.
+// They should only perform visual operations on given data.
+type printer interface {
+	// printStart should print the first message, after the program starts.
+	// This message is printed only once, at the very beginning.
+	printStart(hostname string, port uint16)
+
+	// printProbeSuccess should print a message after each successful probe.
+	// hostname could be empty, meaning it's pinging an address.
+	// streak is the number of successful consecutive probes.
+	printProbeSuccess(hostname, ip string, port uint16, streak uint, rtt float32)
+
+	// printProbeFail should print a message after each failed probe.
+	// hostname could be empty, meaning it's pinging an address.
+	// streak is the number of successful consecutive probes.
+	printProbeFail(hostname, ip string, port uint16, streak uint)
+
+	// printRetryingToResolve should print a message with the hostname
+	// it is trying to resolve an ip for.
+	//
+	// This is only being printed when the -r flag is applied.
+	printRetryingToResolve(hostname string)
+
+	// printTotalDownTime should print a downtime duration.
+	//
+	// This is being called when host was unavailable for some time
+	// but the latest probe was successful (became available).
+	printTotalDownTime(downtime time.Duration)
+
+	// printStatistics should print a message with
+	// helpful statistics information.
+	//
+	// This is being called on exit and when user hits "Enter".
+	printStatistics(s stats)
+
+	// printVersion should print the current version.
+	printVersion()
+
+	// printInfo should a message, which is not directly related
+	// to the pinging and serves as a helpful information.
+	//
+	// Example of such: new version with -u flag.
+	printInfo(format string, args ...any)
+
+	// printError should print an error message.
+	// Printer should also apply \n to the given string, if needed.
+	printError(format string, args ...any)
+}
```

Comment:
Not sure what benefit we are getting by passing these parameters as values instead of a pointer to stats?

I think you have gone this way because it is cleaner and more precise what method uses/needs what? but I'm thinking about all these memory allocations and GC. A few companies that I know of, have tcping running for months/years so the stats struct gets relatively large after a while.

Any thoughts?

Reply:
> I think you have gone this way because it is cleaner and more precise what method uses/needs what?

Yeah, that's the exact reason. So that in the future, if we decide to update to let's say `printStart(hostname string, port uint16)` and add some other info as a 3rd argument, all printers have to be updated, otherwise the code simply won't compile, and compile-time check is way better than any runtime checks.

Yet if we do `printStart(stats *tcpStats)`, nobody knows, except for the printer's author, what info is going to be printed. We assume that the author knows what he's doing, be we can never be sure. Like even now we have some inconsistencies between plane and JSON printers, which is mostly my fault, but that only proves it.
(Also, passing a pointer could potentially lead to side-effects, like changed information on the initial struct, as at was previously)

That's basically my logic behind all the methods, to explicitly tell what info the method is going to print.

Reply:
Valid points. I also would like to keep the code more readable and not prone to errors which your solution works best for it.

Just need to see the impact it could potentially have in the long run. If you think this would be negligible, let me know.

Reply:
Worst case scenario is already there, which is `printStatistics()` as it requires almost every field of `tcpStats`, so we've just passed the entire struct.

If later on we'd need a lot more parameters than we've expected, we can just pass the entire struct again. (and it shouldn't be a big deal to refactor)